src/dag.rs

```diff
@@ -263,47 +263,186 @@ impl<J: jet::Jet> DagLike for Rc<RedeemNode<J>> {
     }
 }
 
+enum Child<D: DagLike> {
+    None,
+    Repeat { idx: usize },
+    New(D),
+}
+
+#[derive(Clone, Debug)]
+enum Previous {
+    /// This is the root element and there are no previous elements
+    Root,
+    /// This is a left child and the previous element is its parent
+    ParentLeft,
+    /// This is a left child and the previous element is its sibling
+    SiblingLeft,
+    /// This is a right child and the previous element is its parent
+    ParentRight,
+}
+
+#[derive(Clone, Debug)]
+struct IterStackItem<D: DagLike> {
+    /// The element on the stack
+    elem: D,
+    /// Whether we have dealt with this item (and pushed its children,
+    /// if any) yet.
+    processed: bool,
+    /// If the item has been processed, the index of its left child, if known
+    left_idx: Option<usize>,
+    /// If the item has been processed, the index of its right child, if known
+    right_idx: Option<usize>,
+    /// Whether the element is a left- or right-child of its parent
+    previous: Previous,
+}
+
+impl<D: DagLike> IterStackItem<D> {
+    /// Constructor for a new stack item with a given element and relationship
+    /// to its parent.
+    fn unprocessed(elem: D, previous: Previous) -> Self {
+        IterStackItem {
+            elem,
+            processed: false,
+            left_idx: None,
+            right_idx: None,
+            previous,
+        }
+    }
+
+    fn left_child(&self, visited: &HashMap<PointerId, usize>) -> Child<D> {
+        match self.elem.left_child() {
+            Some(child) => match visited.get(&PointerId::from(&child)) {
+                Some(&idx) => Child::Repeat { idx },
+                None => Child::New(child),
+            },
+            None => Child::None,
+        }
+    }
+
+    fn right_child(&self, visited: &HashMap<PointerId, usize>) -> Child<D> {
+        match self.elem.right_child() {
+            Some(child) => match visited.get(&PointerId::from(&child)) {
+                Some(&idx) => Child::Repeat { idx },
+                None => Child::New(child),
+            },
+            None => Child::None,
+        }
+    }
+}
+
 /// Iterates over a DAG in _post order_.
 ///
 /// That means nodes are yielded in the order (left child, right child, parent).
 /// Shared nodes appear only once at their leftmost position.
 #[derive(Clone, Debug)]
 pub struct PostOrderIter<D: DagLike> {
-    stack: Vec<D>,
-    maybe_current: Option<D>,
-    visited: HashSet<PointerId>,
+    /// The index of the next item to be yielded
+    index: usize,
+    /// A stack of elements to be yielded; each element is a node, then its left
+    /// and right children (if they exist and if they have been yielded already)
+    stack: Vec<IterStackItem<D>>,
+    visited: HashMap<PointerId, usize>,
```

Comment:
AFAIU, this maps a memory address to its shared index. Please add documentation.